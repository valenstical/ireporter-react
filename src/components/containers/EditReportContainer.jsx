import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import formDataJSON from 'formdata-json';
import { createStructuredSelector } from 'reselect';
import { Redirect } from 'react-router-dom';
import { getUser, getReport } from '../../selectors';
import EditReport from '../views/EditReport';
import fetchReports from '../../actions/getReportAction';
import updateReport from '../../actions/updateReportAction';

/**
 * Container component for the EditReport view
 * @export
 * @class EditReortContainer
 * @extends {Component}
 */
class EditReportContainer extends Component {
  componentDidMount() {
    const { location: { pathname } } = this.props;
    const pathAsArray = pathname.split('/');
    const reportType = pathAsArray[2];
    const reportId = pathAsArray[3];
    const { fetchReports: processfetchReports } = this.props;
    if (reportType && reportId) processfetchReports(`${reportType}/${reportId}`);
  }

  handleUpdateReport = (event) => {
    event.preventDefault();
    const payload = formDataJSON(new FormData(event.target));
    const { updateReport: processUpdateReport } = this.props;
    processUpdateReport(payload);
  }

  render() {
    const { user, report } = this.props;
    if (!user.isLoggedIn) return <Redirect to="/login" />;
    if (report.redirect) return <Redirect to="/404" />;
    return (
      report.data.length > 0
        ? (
          <EditReport
      report={report}
      handleUpdateReport={this.handleUpdateReport}
      />
        )
        : null
    );
  }
}

EditReportContainer.propTypes = {
  user: PropTypes.object.isRequired,
  report: PropTypes.object.isRequired,
  location: PropTypes.object.isRequired,
  fetchReports: PropTypes.func.isRequired,
  updateReport: PropTypes.func.isRequired,
};

const mapStateToProps = createStructuredSelector(
  {
    user: getUser,
    report: getReport
  }
);

const mapDispatchToProps = dispatch => ({
  fetchReports: route => dispatch(fetchReports(route)),
  updateReport: payload => dispatch(updateReport(payload)),

});

export default connect(mapStateToProps, mapDispatchToProps)(EditReportContainer);
