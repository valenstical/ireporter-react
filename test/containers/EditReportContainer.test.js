import React from 'react';
import setup from '../setup';
import EditReportContainer from '../../src/components/containers/EditReportContainer';

const location = {
  pathname: '/path/type/1'
};

describe('<EditReportContainer>', () => {
  it('should render without crashing', () => {
    const wrapper = setup(<EditReportContainer location={location} />);
    expect(wrapper).toMatchSnapshot();
  });
});
