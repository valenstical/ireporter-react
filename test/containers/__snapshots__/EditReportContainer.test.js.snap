// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditReportContainer> should render without crashing 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter
    keyLength={0}
  >
    <Router
      history={
        Object {
          "action": "REPLACE",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/login",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(EditReportContainer)
        location={
          Object {
            "pathname": "/path/type/1",
          }
        }
      >
        <EditReportContainer
          fetchReports={[Function]}
          location={
            Object {
              "pathname": "/path/type/1",
            }
          }
          report={
            Object {
              "data": Array [],
              "isBusy": false,
              "redirect": false,
              "success": false,
            }
          }
          updateReport={[Function]}
          user={
            Object {
              "isBusy": false,
              "message": Array [],
              "success": false,
            }
          }
        >
          <Redirect
            push={false}
            to="/login"
          />
        </EditReportContainer>
      </Connect(EditReportContainer)>
    </Router>
  </BrowserRouter>
</Provider>
`;
